docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -45,44 +45,53 @@ Credit: the content of this section is adapted from Pablo Arias [here](https://p
 
 ## Then why doesn't my code work?
 
-The reason the intro code might break is that when dealing with GPU-compatible classes with virtual functions, there isn't one VTable, but two. The first has the host versions of the virtual functions, while the second has the device functions. We're initializing the class on the host, so it points to the host VTable.
+The reason the intro code might break is that when dealing with GPU-compatible classes with virtual functions, there isn't one VTable, but two. The first holds the host versions of the virtual functions, while the second holds the device functions. 
+
+![VTableDevice](./figures/VirtualFunctions-VTablesHostDevice.png)
+
+Since we construct the class instance on the host, so it's vpointer points to the host VTable.
+
+![VPointerToHost](./figures/VirtualFunctions-VPointerToHost.png)
 
 Our cudaMemcpy faithfully copied all of the members of the class, including the VPointer merrily pointing at host functions, which we then call on the device. 
 
 ## How to fix this
 
-The problem here is that we are initializing the class on the Host. If we were initializing on the Device, we'd get the correct VPointer, and thus the correct functions. In pseudocode, we want to move from
+The problem here is that we are constructing the class on the Host. If we were constructing on the Device, we'd get the correct VPointer, and thus the correct functions (but only for calls on the device). In pseudocode, we want to move from
 
 ```c++
-Instance* hostInstance = new Instance(); // allocate and initialize host
-Instance* deviceInstance; // cudaMalloc'd to allocate
+Base* hostInstance = new Derived(); // allocate and initialize host
+Base* deviceInstance; // cudaMalloc'd to allocate
 cudaMemcpy(deviceInstance, hostInstance); // to initialize the deivce
 Kokkos::parallel_for(... {
   // use deviceInstance
 });
 ```
 
-To one where we initialize on the device using a technique called `placement new`
+To one where we construct on the device using a technique called `placement new`
 
 ```c++
-Instance* deviceInstance; // cudaMalloc'd to allocate it
+Base* deviceInstance; // cudaMalloc'd to allocate it
 Kokkos::parallel_for(... {
-  new((Instance*)deviceInstance) Instance(); // initialize an instance, and place the result in the pointer deviceInstance
+  new((Derived*)deviceInstance) Derived(); // construct an instance in the place, the pointer deviceInstance points to
 });
 ```
 
-This code is extremely ugly, but leads to a properly initialized instance of the class. Note that like with other uses of `new`, you need to later `free` the memory
+This code is extremely ugly, but leads to functional virtual function calls on the device. The vpointer now points to the device Vtable.
```

Comment:
Looks like there are mixed forms of capitalizations used for `vtable` in these changes (`vtable`/`Vtable`/`VTable`). Just for aesthetics, I suggest making them consistent.

Similar issue with `vpointer`.

Reply:
done. It should now be `Vpointer` and `Vtable` everywhere.